docs/container-vm-image-wrapper.md

```diff
@@ -0,0 +1,238 @@
+storage. *Pending introduction and implementation of VMGroup object*
+
+## Implementation 
+
+### Concepts
+
+When a container starts, the container's image is pulled from a remote
+container registry. Each container image consists of multiple layers and only
+the layers that are not already present locally are retrieved from the remote
+registry. 
+
+When a container is being initialized, a union mount filesystem is used to
+combine all the container image's layers together to form the container's
+directory tree. When processes in a container modify contents on the filesystem,
+the processes are not modifying the layers provided by the container image, they
+are modifying a layer of the union filesystem that is tracking the delta between
+the active container and the underlying layers provided by the container image.
+
+The delta between the container image and an active container is ephemeral and
+is automatically freed when container is destroyed. Depending on the union
+mount filesystem implementation used, the delta may be tracked at the block
+level or the file level. When tracked at the file level, the first time a file
+is written to causes the entire file to be copied into the delta layer.
+
+These are the concepts we're building upon to provide an ephemeral disk to a VM
+backed backed by a container image.
+
+### High Level Design
+
+**Standardised KubeVirt VM Image Wrapper**
+
+KubeVirt provides a standardized base wrapper container image that serves up a
+user provided VM image as a network disk consumable by Libvirt. This base
+container image does not have a VM image inside of it, it merely contains
+everything necessary for serving up a VM image in a way Libvirt can consume.
+
+**User Workflow**
+
+Users push VM Images into the container registry using the KubeVirt base
+wrapper container image. By default, the base wrapper container KubeVirt
+provides will serve up any VM image placed in the /vmimage directory as a block
+device consumable by libvirt.
+
+Example:
+```
+cat << END > Dockerfile 
+FROM kubevirt.io:vmimage
+ADD fedora25.qcow2 /vmimage
+END
+docker build -t vmimages/fedora25:latest .
+docker push vmimages/fedora25:latest
+```
+*NOTE: there's more info on how qcow2 is served as a block device later*
+
+Users start a VM using a "container registry backed VM image" by referencing the
+image name they pushed to the registry. Any number of VMs can be started
+anywhere in the cluster referencing the same image.
+
+Example:
+```
+kind: VM
+spec:
+  domain:
+    devices:
+      disks:
+      - type: ContainerImageWrapper:v1
+        - source:
+          name: vmimages/fedora25:latest
+        - target:
+          device: sda
+```
+
+**KubeVirt Runtime Implementation**
+
+When virt-controller sees a VM object with a disk type of
+**ContainerImageWrapper:v1**, virt-controller places the image wrapper container
+in the virt-launcher pod along with the container that monitors the VM process.
```

Comment:
Ah, I see.

The rewrite of the disk 'source' works in this proposal because the iscsi address exposed by the wrapper containers will be predictable as part of how the ContainerImageWrapper:v1 protocol is defined.

When we get to persistent disks where we copy a vm image stored in the container registry to some PVC... we'll have to revisit this and come up with ContainerImageWrapper:**v2** to express how to define the domain xml.   I was punting on that for this proposal.

For the sake of argument, I think we coordinate the ContainerImageWrapper:**v2** functionality without an http endpoint. We could centralize all of this into **virt-controller** and have the controller coordinate provisioning the PVC with the VM image and doing the disk conversion re-write in the VM spec to point to the PVC before virt-handler even sees it.